### PR TITLE
Bunch of new features

### PR DESCRIPTION
**Pull request by EugeneElkin**

### CODE REVIEW

File: src/visual.ts [hunk 1]
```diff
@@ -1362,17 +1446,79 @@ module powerbi.extensibility.visual {
             this.root = null;
         }
 
+        private findDatPointCloserToInterval(
+            findMax: boolean,
+            limit: number,
+            settingVal: number,
+            interval: number
+        ): number {
+            while (findMax ? limit > settingVal && settingVal <= limit - interval : limit < settingVal && settingVal >= limit + interval) {
```
Comment (uve) [inline src/visual.ts]: Please simplify this condition, it's hard to understand how it works.

File: src/visual.ts [hunk 2]
```diff
@@ -733,13 +760,19 @@ module powerbi.extensibility.visual {
...
             this.createScales();
 
             this.applySelectionStateToData();
 
             this.render();
+
+            // The second rendering is necessary to calculate the width of rendered columns with counting of offset (when start and end x axis is set)
+            if (isXaxisBordersOverriden) {
+                this.updateWidthOfColumn(UpdateColumnsWidthMode.calculatePointsDiff);
+                this.render();
```
Comment (uve) [inline src/visual.ts]: Do we need to call render method twice?

File: src/visual.ts [hunk 3]
```diff
@@ -837,13 +871,63 @@ module powerbi.extensibility.visual {
...
+            switch (updateMode) {
+                case UpdateColumnsWidthMode.markOnly :
+                    widthOfColumn = Histogram.PointMarkSize;
+                break;
+                case UpdateColumnsWidthMode.standardCalculation :
+                    widthOfColumn = countOfValues ? (this.viewportIn.width / countOfValues - Histogram.ColumnPadding) : Histogram.MinViewportInSize;
+                break;
+                case UpdateColumnsWidthMode.calculatePointsDiff :
+                default :
+                    // If number of points (bins) is 1, it means that we need to calculate difference between borders (else case)
```
Comment (uve) [inline src/visual.ts]: Move this case to separate function, e.g:
```
default:
  widthOfColumn = this.calculateDefaultWidth(); 
  break;
```

File: src/visual.ts [hunk 4]
```diff
@@ -837,13 +871,63 @@ module powerbi.extensibility.visual {
...
+                    if (countOfValues && countOfValues > 1) {
+                        columnsRect = d3.selectAll("g")[0].filter(function(d: any) { return d.classList.contains("columns"); })[0] as any;
```
Comment (uve) [inline src/visual.ts]: If you are searching for the first item only, use d3.select("g") instead of d3.selectAll("g")[0]

File: src/visual.ts [hunk 5]
```diff
@@ -837,13 +871,63 @@ module powerbi.extensibility.visual {
...
+                        widthOfColumn = columnsRect.children[1].attributes.x.value - columnsRect.children[0].attributes.x.value - Histogram.ColumnPadding;
+                    }
+                    else {
+                        // Find border ticks positions and their offsets to solve issue with 1 bin width
+                        xAxisTicks = $("g.xAxis g.tick text");
```
Comment (uve) [inline src/visual.ts]: Use search inside of parent element only, not in whole element.

File: src/visual.ts (hunk 2 above)
Comment (EugeneElkin) [reply src/visual.ts]: I haven't found another approach for a while. However, I can spend much time to look for another one.

File: src/visual.ts (hunk 3 above)
Comment (EugeneElkin) [reply src/visual.ts]: I will do

File: src/visual.ts (hunk 4 above)
Comment (EugeneElkin) [reply src/visual.ts]: Ok. Thanks!

File: src/visual.ts (hunk 5 above)
Comment (EugeneElkin) [reply src/visual.ts]: +

File: src/visual.ts (hunk 1 above)
Comment (EugeneElkin) [reply src/visual.ts]: Got it!

File: src/visual.ts (hunk 2 above)
Comment (EugeneElkin) [reply src/visual.ts]: I found a new approach an will send it soon.

**Pull request merged**